[ACM-6261]: Restructuring kubevirt and AWS files to reuse 'Accessing a hosted cluster' (#5335)
open-cluster-management/rhacm-docs: clusters/hosted_control_planes/create_hosted_clusters_kubevirt_default_ingress_dns.adoc

[#create-hosted-clusters-kubevirt-default-ingress-dns]
= Default Ingress and DNS behavior

> Every {ocp-short} cluster includes a default application Ingress controller, which must have an wildcard DNS record associated with it. By default, guest clusters that are created by using the HyperShift KubeVirt provider automatically become a subdomain of the underlying {ocp-short} cluster that the KubeVirt virtual machines run on.

Reviewer: ```suggestion
Every {ocp-short} cluster includes a default application Ingress controller, which must have an wildcard DNS record associated with it. By default, guest clusters that are created by using the HyperShift KubeVirt provider automatically become a subdomain of the {ocp-short} cluster that the KubeVirt virtual machines run on.
```

For example, your {ocp-short} cluster might have the following default Ingress DNS entry:

[source,bash]
----
*.apps.mgmt-cluster.example.com
----

> As a result, a KubeVirt guest cluster that is named `guest` and that runs on that underlying {ocp-short} cluster has the following default Ingress:

Reviewer: Is a guest cluster a type of KubeVirt cluster? 

```suggestion
As a result, a KubeVirt cluster that is named `guest` and that runs on the {ocp-short} cluster has the following default Ingress:
```

Author: I don't think a guest cluster is a specific type of KubeVirt cluster, but I will check with engineering.

Author: Engineering confirmed that we can use "a KubeVirt hosted cluster" here instead of "a KubeVirt guest cluster" to mean any OCP cluster that is hosted by KubeVirt VMs. I'll make that change in this file and throughout the KubeVirt docs for consistency.

[source,bash]
----
*.apps.guest.apps.mgmt-cluster.example.com
----

> *Note:* For the default Ingress DNS to work properly, the underlying cluster that hosts the KubeVirt virtual machines must allow wildcard DNS routes. You can configure this behavior by entering the following command: `oc patch ingresscontroller -n openshift-ingress-operator default --type=json -p '[{ "op": "add", "path": "/spec/routeAdmission", "value": {wildcardPolicy: "WildcardsAllowed"}}]'`

Reviewer: ```suggestion
*Note:* For the default Ingress DNS to work properly, the cluster that hosts the KubeVirt virtual machines must allow wildcard DNS routes. You can configure this behavior by entering the following command: `oc patch ingresscontroller -n openshift-ingress-operator default --type=json -p '[{ "op": "add", "path": "/spec/routeAdmission", "value": {wildcardPolicy: "WildcardsAllowed"}}]'`
```

[#create-hosted-clusters-kubevirt-customized-ingress-dns]
== Customizing Ingress and DNS behavior

> If you do not want to use the default Ingress and DNS behavior, you can configure a KubeVirt guest cluster with a unique base domain at creation time. This option requires manual configuration steps during creation and involves three main steps: cluster creation, load balancer creation, and wildcard DNS configuration.

Reviewer: Is it necessary to call out those specific "main" tasks? 
```suggestion
If you do not want to use the default Ingress and DNS behavior, you can configure a KubeVirt guest cluster with a unique base domain at creation time. For this option, you must manually complete the configuration steps during creation and involves three main steps: cluster creation, load balancer creation, and wildcard DNS configuration.
```

Author: If I recall correctly from when we added this content, the engineer wanted me to highlight those specific tasks.

----

+
See the following example output:

+
----
NAME                                       VERSION   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
console                                    4.12.7    False       False         False      30m     RouteHealthAvailable: failed to GET route (https://console-openshift-console.apps.example.hypershift.lab): Get "https://console-openshift-console.apps.example.hypershift.lab": dial tcp: lookup console-openshift-console.apps.example.hypershift.lab on 172.31.0.10:53: no such host
.
.
.
ingress                                    4.12.7    True        False         True       28m     The "default" ingress controller reports Degraded=True: DegradedConditions: One or more other status conditions indicate a degraded state: CanaryChecksSucceeding=False (CanaryChecksRepetitiveFailures: Canary route checks for the default ingress controller are failing)
----

+
The next steps fixes the errors in the output.
+
*Note:* If your cluster is on bare metal, you might need MetalLB so that you can set up load balancer services. For more information, see _Optional: Configuring MetalLB_.

[#set-up-load-balancer]
=== Setting up the load balancer 

> Set up the load balancer that routes to the KubeVirt VMs and assign a wildcard DNS entry to the load balancer IP address. You need to create a load balancer service that routes Ingress traffic to the KubeVirt VMs. A `NodePort` service that exposes the hosted cluster Ingress already exists, so you can export the node ports and create the load balancer service that targets those ports.

Reviewer: Strong description of what the task is and why it's important to the user

. Export the external IP by entering the following command:

+
----
export EXTERNAL_IP=$(oc -n clusters-$CLUSTER_NAME get service $CLUSTER_NAME-apps -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
----

. Configure a wildcard DNS entry that references the IP that is stored in the `$EXTERNAL_IP` path. View the following example DNS entry:

+
[source,bash]
----
*.apps.<hosted-cluster-name\>.<base-domain\>.
----

+
The DNS entry must be able to route inside and outside of the cluster. If you use the example input from step 1, for the cluster that has an external IP value of `192.168.20.30`, the DNS resolutions look like this example:

+
----
dig +short test.apps.example.hypershift.lab

192.168.20.30
----

. Check the hosted cluster status and ensure that it has moved from `Partial` to `Completed` by entering the following command:

+
----
oc get --namespace clusters hostedclusters
----

+
See the following example output:

+
----
NAME            VERSION   KUBECONFIG                       PROGRESS    AVAILABLE   PROGRESSING   MESSAGE
example         4.12.7    example-admin-kubeconfig         Completed   True        False         The hosted control plane is available

> ----

Reviewer: I recommend that you add the links back to the top of the page and then to the landing page that brought the user here.

Author: @dockerymick Is there standard syntax that the RHACM docs use to include a "Back to top" link? Or do we just essentially code an anchor link back to the top of the file?

Reviewer: We do not have a standard, but I think that is fair to add to our internal docs. I usually say, "Return to the beginning of this topic, <<title-of-the-page,Title of the page>>."

Author: Perfect - thanks! I'll follow your lead :)

----

. Access the cluster by entering the following commands:

+
----
hcp create kubeconfig --name $CLUSTER_NAME > $CLUSTER_NAME-kubeconfig
----

+
----
oc --kubeconfig $CLUSTER_NAME-kubeconfig get co
----

+
See the following example output:

+
----
NAME                                       VERSION   AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
console                                    4.12.7    False       False         False      30m     RouteHealthAvailable: failed to GET route (https://console-openshift-console.apps.example.hypershift.lab): Get "https://console-openshift-console.apps.example.hypershift.lab": dial tcp: lookup console-openshift-console.apps.example.hypershift.lab on 172.31.0.10:53: no such host
.
.
.
ingress                                    4.12.7    True        False         True       28m     The "default" ingress controller reports Degraded=True: DegradedConditions: One or more other status conditions indicate a degraded state: CanaryChecksSucceeding=False (CanaryChecksRepetitiveFailures: Canary route checks for the default ingress controller are failing)
----

+
The next steps fixes the errors in the output.
+

> *Note:* If your cluster is on bare metal, you might need MetalLB so that you can set up load balancer services. For more information, see _Optional: Configuring MetalLB_.

Reviewer: I saw that this link can be accessed from the Managing hosted control plane clusters on OpenShift Virtualization doc. I recommend that you add the link for Optional: Configuring MetalLB to the Additional resources section. Keep it italicized here, but add the reference link in the section

Author: Will do!
